dialogue-serde/src/main/java/com/palantir/conjure/java/dialogue/serde/EndpointErrorDecoder.java

```diff
@@ -0,0 +1,250 @@
+            log.debug("Received an error response", diagnosticArgs(response));
+        }
+        try {
+            return decodeInternal(response);
+        } catch (Exception e) {
+            e.addSuppressed(diagnostic(response));
+            throw e;
+        }
+    }
+
+    Optional<RuntimeException> checkCode(Response response) {
+        int code = response.code();
+        switch (code) {
+            case 308:
+                Optional<String> location = response.getFirstHeader(HttpHeaders.LOCATION);
+                if (location.isPresent()) {
+                    String locationHeader = location.get();
+                    try {
+                        UnknownRemoteException remoteException = new UnknownRemoteException(code, "");
+                        remoteException.initCause(
+                                QosException.retryOther(qosReason(response), new URL(locationHeader)));
+                        return Optional.of(remoteException);
+                    } catch (MalformedURLException e) {
+                        log.error(
+                                "Failed to parse location header for QosException.RetryOther",
+                                UnsafeArg.of("locationHeader", locationHeader),
+                                e);
+                    }
+                } else {
+                    log.error("Retrieved HTTP status code 308 without Location header, cannot perform "
+                            + "redirect. This appears to be a server-side protocol violation.");
+                }
+                break;
+            case 429:
+                return Optional.of(response.getFirstHeader(HttpHeaders.RETRY_AFTER)
+                        .map(Longs::tryParse)
+                        .map(Duration::ofSeconds)
+                        .map(duration -> QosException.throttle(qosReason(response), duration))
+                        .orElseGet(() -> QosException.throttle(qosReason(response))));
+            case 503:
+                return Optional.of(QosException.unavailable(qosReason(response)));
+        }
+        return Optional.empty();
+    }
+
+    private T decodeInternal(Response response) {
+        Optional<RuntimeException> maybeQosException = checkCode(response);
+        if (maybeQosException.isPresent()) {
+            throw maybeQosException.get();
+        }
+        int code = response.code();
+        String body;
+        try {
+            body = toString(response.body());
+        } catch (NullPointerException | IOException e) {
+            UnknownRemoteException exception = new UnknownRemoteException(code, "<unparseable>");
+            exception.initCause(e);
+            throw exception;
+        }
+
+        Optional<String> contentType = response.getFirstHeader(HttpHeaders.CONTENT_TYPE);
+        String jsonContentType = "application/json";
+        if (contentType.isPresent() && Encodings.matchesContentType(jsonContentType, contentType.get())) {
```

Comment:
Perhaps we update this to leverage the `Encoding` with something along these lines?
```java
Encodings.matchesContentType(JSON_ENCODING.getContentType(), contentType.get())
```

If it's not terribly troublesome, we may want to take the existing json encoding from the configured set of encodings when it's present, and create a new one as a fallback. This way if the client is configured with encodings wrapped to produce metrics/etc, we can retain that functionality.